Break circular dependency: init -> * -> init by extracting shutdown.h

Empact opened the pull request:
Most includers just wanted to react to pending shutdown.

This isolates access to `fRequestShutdown` and limits access to the shutdown api functions, including the new `CancelShutdown` for setting it to `false`.

Empact commented:
Prompted by #13228

fanquake commented:
Failures in bench_bitcoin in a few of the Travis builds:
```
  CXXLD    bench/bench_bitcoin
libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o): In function `CWalletTx::RelayWalletTransaction(CConnman*)':
wallet.cpp:(.text+0x5cea): undefined reference to `CConnman::NodeFullyConnected(CNode const*)'
collect2: error: ld returned 1 exit status
make[2]: *** [bench/bench_bitcoin] Error 1
```

laanwj commented:
I entertained the idea of creating a `shutdown.(cpp|h)` a few times as well. It seems we have too much non-initialization-related things in init. Concept ACK.

Empact commented:
Based this on #13239, which I expect will address the linker issues.

jonasschnelli commented:
utACK f7033bf5767e674abfee2bd53cd6b9f5c3831482

practicalswift commented:
utACK f7033bf5767e674abfee2bd53cd6b9f5c3831482

promag commented on src/shutdown.cpp:
```diff
@@ -0,0 +1,21 @@
+
+#include <atomic>
+
+std::atomic<bool> fRequestShutdown(false);
```
```cpp
static std::atomic<bool> fRequestShutdown{false};
```

promag commented on src/shutdown.h:
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2017 The Bitcoin Core developers
```
2018.

promag commented on src/txdb.cpp:
```diff
@@ -12,7 +12,7 @@
 #include <util.h>
 #include <ui_interface.h>
-#include <init.h>
+#include <shutdown.h>
```
Keep sorted.

MarcoFalke replied:
Note that you can install `clang-format` and run the https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy script which fixes all white space and sorting issues

promag commented on src/shutdown.cpp:
```diff
@@ -0,0 +1,21 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <atomic>
```
nit, also `#include <shutdown.h>`?

promag reviewed (commented):
Concept ACK.

The PR suggests that the following dependencies wouldn't exist:
```
Circular dependency: init -> net_processing -> init
Circular dependency: init -> validationinterface -> init
```
~~Also, I've removed the first 3 commits locally and it still compiles and reports the same dependencies as above. I wonder why are those commits included here.~~

Edit: oh this is rebased on #13239.

Empact committed:
Break circular dependency: init -> * -> init by extracting shutdown.h

Most includers just wanted to react to pending shutdown.

This isolates access to `fRequestShutdown` and limits access to the shutdown
api functions, including the new `AbortShutdown` for setting it to `false`.

Note I originally called `AbortShutdown` `CancelShutdown` but that name was
already taken by winuser.h
https://travis-ci.org/bitcoin/bitcoin/jobs/386913329

This change also triggered a build error in bench. Fixing it required moving LIBBITCOIN_SERVER after LIBBITCOIN_WALLET in bench_bench_bitcoin_LDADD To make
server definitions in src/net.cpp available to wallet methods in
src/wallet/wallet.cpp. Specifically, solving:

  libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o): In function `CWalletTx::RelayWalletTransaction(CConnman*)':
  wallet.cpp:(.text+0x3f0e): undefined reference to `CConnman::NodeFullyConnected(CNode const*)'
  collect2: error: ld returned 1 exit status

https://travis-ci.org/bitcoin/bitcoin/jobs/392133581

Need for remaining init.h includes confirmed via a thorough search with a more
specific regex:
  \bInterrupt\(\)|\bShutdown\(\)|\bInitLogging\(\)|\bInitParameterInteraction\(\)|\bAppInitBasicSetup\(\)|\bAppInitParameterInteraction\(\)|\bAppInitSanityChecks\(\)|\bAppInitLockDataDirectory\(\)|\bAppInitMain\(\)|\bSetupServerArgs\(\)|\bLicenseInfo\(\)|g_wallet_init_interface|init.h

Empact commented:
* Declared `fRequestShutdown` static
* Included shutdown.h in shutdown.cpp
* Updated copyright years
* Ran `clang-format` to sort includes

Thanks @promag / @MarcoFalke

promag commented:
Still has the following circular dependencies:
```
Circular dependency: init -> net_processing -> init
Circular dependency: init -> validationinterface -> init
```
Should these be removed in this PR?

Empact commented:
@promag Added a commit to remove unused includes of init from 4 files. I'd looked into those before but left them out to narrow this PR.

promag commented:
@Empact needs rebase.

Empact commented:
Rebased

Empact commented:
@jonasschnelli @practicalswift would you take a look at #13239? It's the necessary precondition for this.

Empact commented:
Rebased for #13191 - note this is WIP because I'm working on a Makefile-based solution to the `ld` failures (which unfortunately do not exist on mac), as an alternative to #13239.

Empact commented:
@sipa yeah what I need to do is install linux so that I can repro the build failures locally (unfortunately the only system they're not exhibiting on is Mac). The last commit 665bf42 is a half-finished effort to fix the build with feedback via CI but it would progress a lot faster if I had local failures to run against.

I'll work on it a bit more tonight, in the same style, but failing that I'll have to set up dual-boot or the like.

Empact commented:
I'm going to try again at a more focused intervention.

promag commented:
> @sipa yeah what I need to do is install linux so that I can repro the build failures locally 

@Empact you could use docker instead.

Empact commented:
@promag Good point! I think I'm close to reproing it on mac - the change I made in 44bda43 inverts the test failures, so that the mac build is failing now.
https://travis-ci.org/bitcoin/bitcoin/builds/392143625?utm_source=github_status&utm_medium=notification

Empact commented:
Scratch that! All green.

I'm going to have to run through CI a few more times though to determine which commit to squash the build fix into.

Empact commented:
Ok, looks like an intermittent failure in 69b0e8a, ending in:
```
make[1]: Entering directory '/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-apple-darwin11'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-apple-darwin11'
make -C src qt/bitcoin-qt
make[1]: Entering directory '/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-apple-darwin11/src'
make[1]: Leaving directory '/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-apple-darwin11/src'
/bin/mkdir -p Bitcoin-Qt.app/Contents/MacOS
STRIPPROG="/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../native/bin/x86_64-apple-darwin11-strip" /bin/bash /home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-apple-darwin11/build-aux/install-sh -c -s  src/qt/bitcoin-qt Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt
INSTALLNAMETOOL=/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../native/bin/x86_64-apple-darwin11-install_name_tool  OTOOL=/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../native/bin/x86_64-apple-darwin11-otool STRIP=/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../native/bin/x86_64-apple-darwin11-strip /usr/bin/python3.6 ./contrib/macdeploy/macdeployqtplus Bitcoin-Qt.app -translations-dir=/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../translations -add-qt-tr da,de,es,hu,ru,uk,zh_CN,zh_TW -verbose 2
+ Copying source bundle +
+ Deploying frameworks +
Warning: Could not find any external frameworks to deploy in dist/Bitcoin-Qt.app.
+ Installing qt.conf +
+ Adding Qt translations +
+ Done +
Warning: Could not detect Qt's path, skipping plugin deployment!
/bin/mkdir -p dist/.background
/usr/bin/tiffcp -c none dpi36.background.tiff dpi72.background.tiff dist/.background/background.tiff
/usr/bin/python3.6 contrib/macdeploy/custom_dsstore.py "dist/.DS_Store" "Bitcoin-Core"
/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../native/bin/genisoimage -no-cache-inodes -D -l -probe -V "Bitcoin-Core" -no-pad -r -dir-mode 0755 -apple -o Bitcoin-Core.dmg dist
/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../native/bin/genisoimage: Warning: assuming PC Exchange cluster size of 512 bytes
 30.79% done, estimate finish Thu Jun 14 09:33:16 2018
 61.60% done, estimate finish Thu Jun 14 09:33:16 2018
 92.30% done, estimate finish Thu Jun 14 09:33:16 2018
Total translation table size: 0
Total rockridge attributes bytes: 3197
Total directory bytes: 13084
Path table size(bytes): 118
Max brk space used 0
16257 extents written (31 MB)
```
https://travis-ci.org/bitcoin/bitcoin/jobs/392160315

Seems like it's in master, given this is also failing:
https://travis-ci.org/bitcoin/bitcoin/jobs/392060921

I don't know if 57863ab is an actual fix, or just that the problem is intermittent / unrelated and just hasn't presented itself yet.

Empact commented:
Failure in 63cebe7 confirms that 57863ab is not a fix. Dropping and squashing.

Empact committed:
Drop unused init.h includes

These were entirely unused, as based on successful compilation
and a grep for:
  \bStartShutdown\(\)|\bShutdownRequested\(\)|\bInterrupt\(\)|\bShutdown\(\)|\bInitLogging\(\)|\bInitParameterInteraction\(\)|\bAppInitBasicSetup\(\)|\bAppInitParameterInteraction\(\)|\bAppInitSanityChecks\(\)|\bAppInitLockDataDirectory\(\)|\bAppInitMain\(\)|\bSetupServerArgs\(\)|\bLicenseInfo\(\)|g_wallet_init_interface|init.h

Empact commented:
Ok, I think this is ready for re-review. @sipa thanks for the instigation.

circular-dependencies.py output before, with removals in after indicated with strikethrough:
```
src$ ../contrib/devtools/circular-dependencies.py {*,*/*,*/*/*}.{h,cpp}
Circular dependency: chainparamsbase -> util -> chainparamsbase
Circular dependency: checkpoints -> validation -> checkpoints
~~Circular dependency: index/txindex -> init -> index/txindex~~
Circular dependency: index/txindex -> validation -> index/txindex
~~Circular dependency: init -> net_processing -> init~~
~~Circular dependency: init -> rpc/server -> init~~
~~Circular dependency: init -> txdb -> init~~
~~Circular dependency: init -> validation -> init~~
~~Circular dependency: init -> validationinterface -> init~~
Circular dependency: policy/fees -> txmempool -> policy/fees
Circular dependency: policy/policy -> validation -> policy/policy
Circular dependency: qt/addresstablemodel -> qt/walletmodel -> qt/addresstablemodel
Circular dependency: qt/bantablemodel -> qt/clientmodel -> qt/bantablemodel
Circular dependency: qt/bitcoingui -> qt/utilitydialog -> qt/bitcoingui
Circular dependency: qt/bitcoingui -> qt/walletframe -> qt/bitcoingui
Circular dependency: qt/bitcoingui -> qt/walletview -> qt/bitcoingui
Circular dependency: qt/clientmodel -> qt/peertablemodel -> qt/clientmodel
Circular dependency: qt/paymentserver -> qt/walletmodel -> qt/paymentserver
Circular dependency: qt/recentrequeststablemodel -> qt/walletmodel -> qt/recentrequeststablemodel
Circular dependency: qt/sendcoinsdialog -> qt/walletmodel -> qt/sendcoinsdialog
Circular dependency: qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel
Circular dependency: qt/walletmodel -> qt/walletmodeltransaction -> qt/walletmodel
Circular dependency: rpc/rawtransaction -> wallet/rpcwallet -> rpc/rawtransaction
Circular dependency: txmempool -> validation -> txmempool
Circular dependency: validation -> validationinterface -> validation
Circular dependency: wallet/coincontrol -> wallet/wallet -> wallet/coincontrol
Circular dependency: wallet/fees -> wallet/wallet -> wallet/fees
Circular dependency: wallet/rpcwallet -> wallet/wallet -> wallet/rpcwallet
Circular dependency: wallet/wallet -> wallet/walletdb -> wallet/wallet
~~Circular dependency: index/txindex -> init -> rpc/blockchain -> index/txindex~~
Circular dependency: policy/fees -> policy/policy -> validation -> policy/fees
Circular dependency: policy/rbf -> txmempool -> validation -> policy/rbf
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/addressbookpage
Circular dependency: qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/guiutil
Circular dependency: txmempool -> validation -> validationinterface -> txmempool
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/receivecoinsdialog -> qt/addressbookpage
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/signverifymessagedialog -> qt/addressbookpage
Circular dependency: qt/guiutil -> qt/walletmodel -> qt/optionsmodel -> qt/intro -> qt/guiutil
Circular dependency: qt/addressbookpage -> qt/bitcoingui -> qt/walletview -> qt/sendcoinsdialog -> qt/sendcoinsentry -> qt/addressbookpage
```

There's an intermittent failure but as described above, it seems to be an unrelated Qt path problem.

ken2812221 reviewed (approved):
utACK ab30690763b9396abdf2b44f7ef1077ad7ffd29e

Empact commented:
@jonasschnelli @practicalswift @promag @sipa take another look?

ken2812221 commented:
utACK 1fabd59e7e870fae73bfbcfb227dd7452de94726

sipa commented:
@Empact In your first commit you write "These were entirely unused, as based on successful compilation". That's not a sufficient argument - init.h could still be included indirectly through another header, while exported symbols from init.h are still used.

This doesn't mean it wrong to remove them, just that "it compiles without them" isn't a sufficient reason. Tools like `iwyu` can help with this.

Empact commented:
@sipa thanks I’ll check that out

Empact commented:
Oh btw I did manually check using this grep:

```
  \bStartShutdown\(\)|\bShutdownRequested\(\)|\bInterrupt\(\)|\bShutdown\(\)|\bInitLogging\(\)|\bInitParameterInteraction\(\)|\bAppInitBasicSetup\(\)|\bAppInitParameterInteraction\(\)|\bAppInitSanityChecks\(\)|\bAppInitLockDataDirectory\(\)|\bAppInitMain\(\)|\bSetupServerArgs\(\)|\bLicenseInfo\(\)|g_wallet_init_interface|init.h
```

But I'll check again with the above and using `iwyu`, which is probably more robust.

sipa commented:
utACK 1fabd59e7e870fae73bfbcfb227dd7452de94726

Empact commented:
@practicalswift @jonasschnelli @promag could I get another look?

Alternatively I could split this into 2 PRS, the removal of the unnecessary includes and the shutdown.h extraction, if that would make review easier.
